### DISCUSSION TITLE
How to paginate with only a "page" parameter?

### DISCUSSION THREAD

**Discussion by GollyJer** [urql-graphql/urql#1719] [Q&A] [answered]
Here is a sample query to an endpoint I'm consuming and cannot change.

```javascript 
{
  transfers(page: 1, recipient: "tester1", sender: "tester2") {
    id
    created_at_time
    assets {
      id
    }
  }
}
```

I have created the following hook to "page" and get all the results.
```javascript
export const useTransfers = ({
  limit = 100,
  recipient,
  sender,
}) => {
  const page = useRef(1);

  const [query] = useQuery({
    query: GET_TRANSFERS,
    pause,
    variables: {
      page: page.current,
      recipient,
      sender,
      limit,
    },
  });
  if (query.error) console.log('useTransfers ERROR\n', query.error, '\n'); // eslint-disable-line
  const queryResults = query.data?.transfers;

  const prevResultCount = usePrevious(queryResults?.length);
  const isNewResultCount = queryResults?.length !== prevResultCount;
  const resultHitLimit = queryResults?.length === limit;

  useEffect(() => {
      if (isNewResult || resultHitLimit) {
        page.current += 1;
        setAggregateResults((prevResults) => [...prevResults, ...queryResults]);
      }
  }, [queryResults]); //eslint-disable-line

const noPagesLeft = page.current > 1 && queryResults?.length === 0;
  if (!pause && noPagesLeft) {
    // deduping here because we have no control when paging with cache-and-network.
    const dedupedResults = [
      ...aggregateResults.reduce((map, obj) => map.set(obj.id, obj), new Map()).values(),
    ];
    return [dedupedResults, true];
  }
  return [[], false];
}
```
This works but is the result of an imbecile hacker. 
The Pagination urql doc says "Given we have a schema that uses some form of `offset` and `limit` based pagination" which this schema does not.

I have no idea how to approach this correctly and would love your ideas/help. Thanks!

**Comment by kitten** [comment]
This is a tricky one to answer as I'd say _it depends_.

What I'm noting from your implementation is that it may be brittle and based on a lot of assumptions; for instance, your effect assumes that it's safe to modify a ref without getting into issues where React aborts the effect, where the same may apply to `usePrevious`. So there's a couple of assumptions I'd have to make to comment on this.

I could also comment on the GraphQL schema itself, since it's really not well set up to paginate, since it doesn't deliver even some metadata on how many pages there are and such.

That said, I'd always recommend a component-based approach to pagination where possible: https://github.com/FormidableLabs/urql/blob/main/examples/with-pagination/src/PaginatedNpmSearch.jsx
We have this example set up to show what that means. Where possible, I'd opt to split each page up into a separate child that renders just a fragment of a page rather than merging pages.

Where that's not possible, it does get trickier. You're completely right in accummulating results, but you do have to be careful not to introduce too much state. We have some helpers set up for set pagination patterns in our normalized cache, but not for all patterns, as there are too many non-standard ways to implement pagination. That said, I'd remove state wherever possible to ensure that a minimum amount of edge cases is introduced.

**Reply by GollyJer** [reply]
Hi @kitten. Thank you so much for taking the time to respond! Super helpful.

After thinking through your comments this is where it landed. 
The biggest change is all state to refs which can be edited without causing renders.

Code is posted as answer...

**Comment by GollyJer** [comment, marked as answer]
It's a super hack but because we don't control the GraphQL schema it's the best of a bad situation. I'm happy it at least works at this point though. Thanks again! 🎉

```javascript 
export const useTransfers = ({ recipient, sender, shouldPause = false }) => {
  const pause = !recipient || !sender || shouldPause;
  const page = useRef(1);
  const results = useRef([]);

  const [query, refetch] = useQuery({
    query: GET_TRANSFERS,
    pause,
    variables: {
      page: page.current,
      recipient,
      sender,
    },
  });
  if (query.error) console.log('useTransfers ERROR\n', JSON.stringify(query.error), '\n'); // eslint-disable-line

  const queryResults = query.data?.transfers;
  const prevSender = usePrevious(sender);
  if (sender !== prevSender) {
    page.current = 1;
    results.current = [];
  }

  const prevResultCount = usePrevious(queryResults?.length);
  const isNewResultCount = queryResults?.length !== prevResultCount;
  const isNewResult = queryResults?.length > 0 && isNewResultCount;
  const resultHitLimit = queryResults?.length === limit;

  useEffect(() => {
    if (!pause && !query.fetching) {
      if (isNewResult || resultHitLimit) {
        page.current += 1;
        results.current = [...results.current, ...queryResults];
        refetch();
      }
    }
  }, [queryResults]); //eslint-disable-line

  const noPagesLeft = page.current > 1 && queryResults?.length === 0;
  if (!pause && noPagesLeft) {
    // deduping here because we have no control when paging with cache-and-network.
    const dedupedById = [
      ...results.current.reduce((map, obj) => map.set(obj.id, obj), new Map()).values(),
    ];
    return [dedupedById];
  }
  return [[]];
};
```
